cloud/amazon: add option to skip TLS verification for backup/restore

sravotto opened the pull request:
Previously, backup/restore jobs running against secure, on-premises S3 object stores that use self signed certificate would fail, unless cloudstorage.http.custom_ca is set.
However, customers want the ability to bypass the certificate validation in test environments.

To address this, we added a new URL parameter (AWS_SKIP_TLS_VERIFY) that can be set in the backup URL to bypass certificate validation. Note that a CA certificate is still required; this parameter means that the client will not verify the certificate.

Warning: Use this query parameter with caution, as it creates MITM vulnerabilities.

This change also adds a test case in roachtest for verification against a microCeph cluster.

Epic: none

Fixes: 147910

Release note: None

sravotto committed:
cloud/amazon: add option to skip TLS verification for backup/restore

Previously, backup/restore jobs running against secure, on-premises
S3 object stores that use self signed certificate would fail, unless
cloudstorage.http.custom_ca is set.
However, customers want the ability to bypass the certificate validation in
test environments.

To address this, we added a new URL parameter (AWS_SKIP_TLS_VERIFY) that can
be set in the backup URL to bypass certificate validation.
Note that a CA certificate is still required; this parameter means that
the client will not verify the certificate.

Warning: Use this query parameter with caution, as it creates
MITM vulnerabilities.

This change also adds a test case in roachtest for verification against
a microCeph cluster.

Epic: none

Fixes: 147910

Release note: None

msbutler reviewed (commented):
nice! flushing comments on non roachtest pkg code

msbutler commented on pkg/cmd/roachtest/tests/s3_clone_backup_restore.go:
```diff
@@ -24,48 +24,76 @@ import (
+
+const (
+	s3ClonePlain      s3CloneSecureOption = iota // Use HTTP
+	s3CloneRequireTLS                            // Use HTTPS, but skip certification verification.
```
uber nit: you could rename `s3CloneRequireTLS` to `s3CloneTLSWithSkipVerify` to suggest this is non-default, and then `s3CloneVerifyTLS` to `s3CloneTLS`.

msbutler reviewed (commented):
roachtest changes look good too.

sravotto committed:
cloud/amazon: renaming cloud.Config to cloud.HTTPClientConfig.

Renaming cloud.Config to cloud.HTTPClientConfig to be more specific.
Added InjectTestingRetryMaxAttempts to limit number of request retries
during testing.

Release note: None

sravotto reviewed (commented):
Thanks for the review - please take another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147914#-:-OSPifr496NVCGR1aqmc:bvp5iuk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir and @msbutler)



<!-- Sent from Reviewable.io -->

sravotto commented on pkg/cloud/amazon/s3_storage.go:
```diff
@@ -573,24 +593,27 @@ func (s *s3Storage) newClient(ctx context.Context) (s3Client, string, error) {
 
-	// TODO(yevgeniy): Revisit retry logic.  Retrying 10 times seems arbitrary.
-	retryMaxAttempts := 10
-	addLoadOption(config.WithRetryMaxAttempts(retryMaxAttempts))
```
Done.
<!-- Reviewable comment -OSQRPneAFkvEa2quoy8:b-896fix -->
<!-- Sent from Reviewable.io -->

sravotto reviewed (commented):
I added a global variable `retryMaxAttempts` based on your suggestion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147914#-:-OSQU34605AHKJpxoNlY:b-lsu3bj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @BramGruneir and @msbutler)



<!-- Sent from Reviewable.io -->

msbutler reviewed (approved):
nice!

sravotto commented:
bors r+
